src/support/allocators/node_allocator.h

```diff
@@ -0,0 +1,406 @@
+{
+    const auto alignment_max = std::max(std::alignment_of_v<T>, std::alignment_of_v<FreeList>);
+    const auto size_max = std::max(sizeof(T), sizeof(FreeList));
+
+    // find closest multiple of alignment_max that holds size_max
+    return ((size_max + alignment_max - 1U) / alignment_max) * alignment_max;
+}
+
+} // namespace detail
+
+/**
+ * Actually holds and provides memory to an allocator. MemoryResource is an immobile object. It
+ * stores a number of memory pools which are used to quickly give out memory of a fixed allocation
+ * size. The class is purposely kept very simple. It only knows about "Allocations" and "Pools".
+ *
+ * - Pool: MemoryResource allocates one memory pool at a time. These pools are kept around until the
+ * memory resource is destroyed.
+ *
+ * - Allocations: Node-based containers allocate one node at a time. Whenever that happens, the
+ * MemoryResource's Allocate() gives out memory for one node. These are carved out from a previously
+ * allocated memory pool, or from a free list if it contains entries. Whenever a node is given back
+ * with Deallocate(), it is put into that free list.
+ */
+template <size_t ALLOCATION_SIZE_BYTES>
+class MemoryResource
+{
+    /**
+     * Size in bytes to allocate per pool, currently hardcoded to multiple of ALLOCATION_SIZE_BYTES
+     * that comes closest to 256 KiB.
+     */
+    static constexpr size_t POOL_SIZE_BYTES = (262144 / ALLOCATION_SIZE_BYTES) * ALLOCATION_SIZE_BYTES;
+
+public:
+    /**
+     * Construct a new MemoryResource object that uses the specified allocation size to optimize for.
+     */
+    MemoryResource() = default;
+
+    /**
+     * Copying/moving a memory resource is not allowed; it is an immobile object.
+     */
+    MemoryResource(const MemoryResource&) = delete;
+    MemoryResource& operator=(const MemoryResource&) = delete;
+    MemoryResource(MemoryResource&&) = delete;
+    MemoryResource& operator=(MemoryResource&&) = delete;
+
+    /**
+     * Deallocates all allocated pools.
+     *
+     * There's no Clear() method on purpose, because it would be dangerous. E.g. when calling
+     * clear() on an unordered_map, it is not certain that all allocated nodes are given back to the
+     * MemoryResource. Microsoft's STL still uses a control structure that might have the same size
+     * as the nodes, and therefore needs to be kept around until the map is actually destroyed.
+     */
+    ~MemoryResource() = default;
+
+    /**
+     * Allocates memory for n * sizeof(T). Only when n==1 and T's size matches ALLOCATION_SIZE_BYTES
+     * the pool's memory is used. Otherwise, the allocation is forwarded to ::operator new().
+     *
+     * @tparam T Object to allocate memory for.
+     * @param n Number of objects to allocate for
+     */
+    template <typename T>
+    [[nodiscard]] T* Allocate(size_t n)
+    {
+        if constexpr (ALLOCATION_SIZE_BYTES == detail::RequiredAllocationSizeBytes<T>()) {
```

Comment:
Hm, I don't really want to add a counter for `m_free_allocations` but also don't want to iterate through the freelist. I think it's ok to keep the `MemoryResource::DynamicMemoryUsage` as it is because all that memory is the responsibility of that class. I got rid of the `if constexpr` though in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5.